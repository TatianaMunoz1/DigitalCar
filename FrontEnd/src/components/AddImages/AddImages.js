import React, { useState, useEffect } from "react";
import styles from "./AddImages.module.css";
import Swal from "sweetalert2";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faPlus,
  faTimes,
  faArrowUp,
  faArrowDown,
} from "@fortawesome/free-solid-svg-icons";
import Inputs from "../Inputs/Inputs";

function AddImages(props) {
  const [listImages, setListImages] = useState(null);
  const [validation, setValidation] = useState(false);
  const [newCarImages, setNewCarImages] = useState([]);
  const [newCarImage, setNewCarImage] = useState({
    title: "",
    url: "",
    position: null,
  });
  const [imageValidation, setImageValidation] = useState({
    title: false,
    url: false,
  });
  const expressions = {
    title: /^[a-zA-Z0-9_.+-:", ]+$/,
    url: /(^.)/,
  };

  useEffect(() => {
    props.handleAddNewImages([...newCarImages]);
  }, [newCarImages]);

  const handleInputImageChange = (e) => {
    const target = e.target;
    const value = target.value;
    const name = target.name;

    setNewCarImage({
      ...newCarImage,
      [name]: value,
    });
  };
  const handleAddImages = () => {
    if (newCarImage.title === "" || newCarImage.url === "") {
      setValidation(true);
    } else {
      const pos = newCarImages.length + 1;
      setValidation(false);
      setNewCarImages([...newCarImages, { ...newCarImage, position: pos }]);
      setNewCarImage({
        title: "",
        url: "",
        position: null,
      });
    }
  };

  useEffect(() => {
    const handleDeleteImage = (idImage) => {
      Swal.fire({
        title: "¿Desea borrar la imagen?",
        icon: "warning",
        showCancelButton: true,
        background: "#f3f3f3",
        confirmButtonColor: "#b40505",
        cancelButtonColor: "#000000",
        confirmButtonText: "Borrar",
        cancelButtonText: "No borrar",
      }).then((result) => {
        if (result.isConfirmed) {
          const array = [...newCarImages];
          array.splice(idImage, 1);
          setNewCarImages([...array]);
        }
      });
    };

    const handlePositionDown = (idImage) => {
      const newIdImage = parseInt(idImage) + 1;
      const list = [...newCarImages];
      const elementMooved = list.splice(idImage, 1);
      list.splice(newIdImage, 0, ...elementMooved);
      const arrayImages = [];
      list.map((element, index) => {
        arrayImages.push({ ...element, position: index + 1 });
        return arrayImages;
      });
      setNewCarImages([...arrayImages]);
    };

    const handlePositionUp = (idImage) => {
      const newIdImage = idImage - 1;
      const list = [...newCarImages];
      const elementMooved = list.splice(idImage, 1);
      list.splice(newIdImage, 0, ...elementMooved);
      const arrayImages = [];
      list.map((element, index) => {
        arrayImages.push({ ...element, position: index + 1 });
        return arrayImages;
      });
      setNewCarImages([...arrayImages]);
    };

    const listImage = () =>
      newCarImages.map((carImage, index) => {
        return (
          <div
            key={`${carImage.title}_${index}`}
            className={styles.new_image_box}
            id={index}
          >
            <div className={styles.new_image}>
              <input
                className={styles.input}
                type="text"
                value={carImage.url}
                disabled
              />
              <input
                className={styles.input}
                type="text"
                value={carImage.title}
                disabled
              />
            </div>
            <div className={styles.btn_container_new_image}>
              <div className={styles.position}>
                <button
                  className={styles.icon_position_up}
                  type="button"
                  id={index}
                  onClick={(e) => handlePositionUp(e.target.id)}
                >
                  {index === 0 && (
                    <FontAwesomeIcon
                      className={styles.icon_position_disabled}
                      icon={faArrowUp}
                    />
                  )}
                  {index > 0 && (
                    <FontAwesomeIcon
                      className={styles.icon_position}
                      icon={faArrowUp}
                      id={index}
                      onClick={(e) => handlePositionUp(e.target.id)}
                    />
                  )}
                </button>
                <button
                  className={styles.icon_position_down}
                  type="button"
                  id={index}
                  onClick={(e) => handlePositionDown(e.target.id)}
                >
                  {index === newCarImages.length - 1 && (
                    <FontAwesomeIcon
                      className={styles.icon_position_disabled}
                      icon={faArrowDown}
                    />
                  )}
                  {index < newCarImages.length - 1 && (
                    <FontAwesomeIcon
                      className={styles.icon_position}
                      icon={faArrowDown}
                      id={index}
                      onClick={(e) => handlePositionDown(e.target.id)}
                    />
                  )}
                </button>
              </div>
              <button
                className={styles.btn_delete_image}
                type="button"
                id={index}
                onClick={(e) => handleDeleteImage(e)}
              >
                <FontAwesomeIcon icon={faTimes} />
              </button>
            </div>
          </div>
        );
      });
    setListImages(listImage);
  }, [newCarImages]);

  return (
    <div className={styles.images}>
      <div className={styles.new_images}>
        {newCarImages.length > 0 && <div>{listImages}</div>}
      </div>

      <div className={styles.add_images_section}>
        <div className={styles.add_images}>
          <Inputs
            label="URL imágen"
            className={styles.input_image_url}
            onChange={handleInputImageChange}
            value={newCarImage.url}
            name="url"
            type="text"
            id="image_url"
            placeholder="Insertar http://"
            errorLegend="Insertar URL de la imagen http://"
            regularExpression={expressions.url}
            validState={imageValidation}
            setValidState={setImageValidation}
          />
          <Inputs
            label="Título"
            onChange={handleInputImageChange}
            value={newCarImage.title}
            name="title"
            type="text"
            id="image_url"
            placeholder="Titulo imagen"
            errorLegend="Ingrese el titulo de la imagen"
            regularExpression={expressions.title}
            validState={imageValidation}
            setValidState={setImageValidation}
          />
        </div>
        <div className={styles.btn_container}>
          <button
            className={styles.btn_add_image}
            type="button"
            onClick={handleAddImages}
          >
            <FontAwesomeIcon icon={faPlus} />
          </button>
        </div>
      </div>
    </div>
  );
}

export default AddImages;
