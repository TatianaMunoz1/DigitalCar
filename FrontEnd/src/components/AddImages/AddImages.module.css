.add_images_section,
.new_image_box {
  background: #e9e9e9;
  box-shadow: 2px 2px 5px rgba(0, 0, 0, 0.15);
  display: flex;
  width: 100%;
  gap: 10px;
  padding: 20px 10px;
  align-items: center;
}
.new_image_box:hover {
  background: #e0e0e0;
  transition: all 0.3s ease-out;
}
.new_image_box {
  gap: 2px;
  justify-content: center;
  margin-bottom: 12px;
}
.add_images,
.new_image {
  display: grid;
  grid-template-rows: 1fr 1fr;
  gap: 20px;
  width: 100%;
}
.input {
  height: 30px;
  width: 100%;
  padding: 6px;
  font-size: 14px;
  background: #ffffff;
  box-shadow: 0px 1px 5px rgba(0, 0, 0, 0.15);
  border: none;
  border-radius: 5px;
}
.btn_add_image,
.btn_delete_image {
  height: 30px;
  min-width: 30px;
  max-width: 30px;
  border-radius: 100%;
  font-size: 15px;
  line-height: 30px;
  padding: 0px;
  background: #ff7e19;
  color: #fff;
  border: none;
  cursor: pointer;
}
.btn_delete_image {
  font-size: 15px;
  line-height: 25px;
  height: 25px;
  min-width: 25px;
  max-width: 25px;
}
.btn_delete_image {
  background: #000000;
}
.btn_delete_image:hover {
  background: #7c7c7c;
  transition: all 0.3s ease-out;
}
.btn_delete_image:hover {
  background: #7c7c7c;
  transition: all 0.3s ease-out;
}
.btn_add_image:hover {
  background-color: #ffa25a;
  transition: all 0.3s ease-out;
}
.btn_container {
  width: 60px;
  display: flex;
  justify-content: right;
}
.new_images {
  padding: 15px 0px;
}
.position {
  font-size: 20px;
  align-items: center;
  justify-content: space-between;
  display: grid;
  grid-template-rows: 1fr 1fr;
  gap: 5px;
}
.icon_position_up,
.icon_position_down {
  min-width: 30px;
  max-width: 30px;
  height: 30px;
  background: transparent;
  color: #696969;
  border: none;
  font-size: 15px;
  text-align: center;
}
.icon_position {
  cursor: pointer;
  height: 30px;
  width: 100%;
}
.icon_position_disabled {
  color: #c2c2c2;
}
.icon_position:hover {
  color: #ffa25a;
  transition: all 0.3s ease-out;
}
.icon_position:active {
  color: #777777;
  transition: all 0.3s ease-out;
}
.new_images_by_position_container {
  display: none;
}
.btn_container {
  width: 65px;
  display: flex;
  justify-content: right;
  align-self: center;
}
.btn_container_new_image {
  width: 65px;
  height: auto;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

/*-----TABLET-----*/
@media (min-width: 760px) and (max-width: 1280px) {
  .input {
    height: 40px;
  }
  .add_images_section,
  .new_image_box {
    gap: 10px;
    padding: 25px;
  }
  .add_images,
  .new_image {
    gap: 15px;
  }
  .icon_position_up,
  .icon_position_down {
    font-size: 16px;
  }
  .btn_add_image,
  .btn_delete_image {
    height: 40px;
    min-width: 40px;
    max-width: 40px;
    border-radius: 100%;
    font-size: 20px;
    line-height: 40px;
    padding: 0px;
    cursor: pointer;
  }
  .btn_container {
    width: 85px;
    display: flex;
    justify-content: right;
    align-items: center;
  }
  .btn_container_new_image {
    width: 85px;
    height: auto;
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
  .new_images_by_position_container {
    display: contents;
  }
  .new_images_by_position {
    width: 120px;
    height: 120px;
    object-fit: cover;
  }
}

/*-----DESKTOP-----*/
@media (min-width: 1281px) {
  .input {
    height: 40px;
  }
  .add_images,
  .new_image {
    display: grid;
    grid-template-columns: 8fr 2fr;
    grid-template-rows: 1fr;
    height: auto;
    gap: 20px;
    width: 100%;
  }
  .add_images_section,
  .new_image_box {
    padding: 20px 30px;
    gap: 10px;
  }
  .add_images_section {
    padding: 30px;
  }
  .icon_position_up,
  .icon_position_down {
    font-size: 16px;
  }
  .btn_container {
    width: 85px;
    display: flex;
    justify-content: right;
    align-self: flex-end;
  }
  .btn_add_image,
  .btn_delete_image {
    height: 40px;
    min-width: 40px;
    max-width: 40px;
    border-radius: 100%;
    font-size: 20px;
    line-height: 40px;
    padding: 0px;
    align-self: flex-end;
    cursor: pointer;
  }
  .btn_container_new_image {
    width: 85px;
    height: auto;
    display: flex;
    justify-content: space-between;
    align-content: center;
  }
  .btn_delete_image {
    align-self: center;
  }
  .new_images_by_position_container {
    display: contents;
  }
  .new_images_by_position {
    width: 120px;
    height: 120px;
    object-fit: cover;
  }
}
