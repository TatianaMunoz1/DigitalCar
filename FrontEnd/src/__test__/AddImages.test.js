import React, { useEffect } from "react";
import { shallow, mount, render, mock } from "enzyme";
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import AddImages from "../components/AddImages/AddImages.js";

describe("add images", () => {
  const history = createMemoryHistory();
  const wrapper = mount(
    <Router history={history}>
      <AddImages />
    </Router>
  );
  test("must be exist", () => {
    expect(wrapper.text().includes("*Faltan completar campos")).toBeTruthy();
  });
  test("render add images ", () => {
    const wrapper = shallow(<AddImages />);
    expect(wrapper).toMatchSnapshot();
  });

  test("should call handleInputFeatureChange method", () => {
    const onClickSpy = jest.fn();
    const wrapper = shallow(<AddImages handleInputImageChange={onClickSpy} />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find("button").simulate("click");
    expect(onClickSpy).toHaveBeenCalled();
  });
  test("must render icon", () => {
    expect(wrapper.find("FontAwesomeIcon")).toHaveLength(1);
  });
  test("icon button click", () => {
    wrapper.find("FontAwesomeIcon").simulate("click");
  });
});
