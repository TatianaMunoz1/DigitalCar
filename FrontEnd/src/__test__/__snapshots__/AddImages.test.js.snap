// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add images render add images  1`] = `ShallowWrapper {}`;

exports[`add images should call handleInputFeatureChange method 1`] = `ShallowWrapper {}`;
